### PR TITLE
support MRL embeddings for qwen2

### PR DESCRIPTION
**Pull request by magdyksaleh**

### CODE REVIEW

File: server/lorax_server/models/flash_causal_lm.py [hunk 1]
```diff
@@ -887,16 +899,18 @@ def warmup(self, batch: FlashCausalLMBatch, max_new_tokens: int):
...
                 logger.info("Warming up to max_new_tokens: {}", max_new_tokens)
-                with tqdm(total=max_new_tokens, desc="Warmup to max_total_tokens") as pbar:
-                    for _ in range(max_new_tokens):
-                        cur_seqlen = batch.max_seqlen
-                        _, batch = self.generate_token(batch, is_warmup=True)
-                        new_seqlen = batch.max_seqlen
-                        pbar.update(new_seqlen - cur_seqlen)
-                        if new_seqlen >= max_total_tokens - get_speculative_tokens():
-                            break
+                # with tqdm(total=max_new_tokens, desc="Warmup to max_total_tokens") as pbar:
```
Comment (magdyksaleh) [inline server/lorax_server/models/flash_causal_lm.py:906]: todo: fix based on model type

File: server/lorax_server/models/flash_causal_lm.py [hunk 2]
```diff
@@ -887,16 +899,18 @@ def warmup(self, batch: FlashCausalLMBatch, max_new_tokens: int):
                 self.device,
             )
 
+            print("KV cache: ", self.kv_cache)
```
Comment (magdyksaleh) [inline server/lorax_server/models/flash_causal_lm.py:902]: rm

File: server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 3]
```diff
@@ -496,7 +495,64 @@ def forward(
...
+            block_tables,
+            slots,
+            input_lengths,
+            max_s,
+            prefill_cache_indices,
+            adapter_data,
+        )
+        batch_size = hidden_states.shape[0] // max_s
+        hidden_states = hidden_states.reshape(batch_size, max_s, -1)
+        mean_hidden_states = hidden_states.mean(1)
```
Comment (arnavgarg1) [inline server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py:555]: Is this taking the mean along the sequence dimension?

File: server/lorax_server/models/flash_qwen2.py [hunk 4]
```diff
@@ -83,7 +93,20 @@ def __init__(
...
+                    "No embedding_dim was provided so we cannot load the model."
+                    "Please pass in an embedding_dim to the model."
+                )
+
+        self._supports_embeddings = embedding_dim is not None
+        if self._supports_embeddings:
+            model = FlashQwen2ForEmbeddings(config, weights)
+        else:
+            model = FlashQwen2ForCausalLM(config, weights)
+
```
Comment (arnavgarg1) [inline server/lorax_server/models/flash_qwen2.py:109]: Maybe:

```python
# Determine if the model supports embeddings
self._supports_embeddings = not weights.has_tensor('lm_head.weight')

# If the model supports embeddings but no embedding_dim is provided, raise an error
if self._supports_embeddings and embedding_dim is None:
    raise ValueError(
        "Model does not have 'lm_head.weight', so it is presumed to be for embeddings. "
        "No 'embedding_dim' was provided, so we cannot load the model. "
        "Please pass in an 'embedding_dim' to the model."
    )

# Instantiate the appropriate model
if self._supports_embeddings:
    model = FlashQwen2ForEmbeddings(config, weights)
else:
    model = FlashQwen2ForCausalLM(config, weights)
```

File: server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py (hunk 3 above)
Comment (magdyksaleh) [reply server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py:555]: yes

File: server/lorax_server/models/flash_qwen2.py (hunk 4 above)
Comment (magdyksaleh) [reply server/lorax_server/models/flash_qwen2.py:109]: sure

File: server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 5]
```diff
@@ -496,7 +495,64 @@ def forward(
...
+        super().__init__()
+        self.config = config
+
+        self.model = FlashQwen2Model(config, weights)
+        self.max_past = config.sliding_window
+        self.output_weight = weights.get_tensor("linear.weight")
+        self.output_bias = weights.get_tensor("linear.bias")
+        # To satisfy the parent class interface
+        # TODO: fix
+        self.lm_head = None
```
Comment (arnavgarg1) [inline server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py:516]: Shouldn't be needed at all no? What's the parent class here?

Comment (magdyksaleh) [reply server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py:516]: The FlashQwen2 class' parent FlashCasualLm has a call to model.lm_head and so it breaks if this is not defined

File: server/lorax_server/utils/sources/hub.py [hunk 6]
```diff
@@ -28,19 +28,41 @@ def weight_hub_files(
...
+    else:
+        filenames = [
+            s.rfilename
+            for s in info.siblings
+            if s.rfilename.endswith(extension)
+            and len(s.rfilename.split("/")) == 1
+            and "arguments" not in s.rfilename
+            and "args" not in s.rfilename
+            and "training" not in s.rfilename
+        ]
```
Comment (arnavgarg1) [inline server/lorax_server/utils/sources/hub.py:65]: nit: For better readibility

```python
def is_valid_filename(filename: str, max_filename_depth: int = 2) -> bool:
        excluded_terms = ["arguments", "args", "training"]
        return (
            filename.endswith(extension)
            and len(filename.split("/")) <= max_filename_depth
            and not any(term in filename for term in excluded_terms)
        )

def weight_hub_files(
    model_id: str,
    revision: Optional[str] = None,
    extension: str = ".safetensors",
    api_token: Optional[str] = None,
    embedding_dim: Optional[int] = None,
) -> List[str]:
    """Get the weights filenames on the hub"""
    api = get_hub_api(token=api_token)
    info = api.model_info(model_id, revision=revision)

    max_filename_depth = 2 if embedding_dim is not None else 1
    filenames = [
        s.rfilename for s in info.siblings if is_valid_filename(s.rfilename, max_filename_depth)
    ]

    if embedding_dim is not None:
        embedding_tensor_file = f"2_Dense_{embedding_dim}/model{extension}"
        if embedding_tensor_file not in filenames:
            raise ValueError(f"No embedding tensor file found for embedding dim {embedding_dim}")

        filenames = [
            filename for filename in filenames
            if len(filename.split("/")) < 2 or filename == embedding_tensor_file
        ]

    if not filenames:
        raise EntryNotFoundError(
            f"No {extension} weights found for model {model_id} and revision {revision}.",
            None,
        )

    return filenames

File: server/lorax_server/utils/sources/hub.py [hunk 7]
```diff
@@ -28,19 +28,41 @@ def weight_hub_files(
...
+            s.rfilename
+            for s in info.siblings
+            if s.rfilename.endswith(extension)
+            and len(s.rfilename.split("/")) <= 2
+            and "arguments" not in s.rfilename
+            and "args" not in s.rfilename
+            and "training" not in s.rfilename
+        ]
+        # Only include the layer for the correct embedding dim
+        embedding_tensor_file = f"2_Dense_{embedding_dim}/model.safetensors"
```
Comment (arnavgarg1) [inline server/lorax_server/utils/sources/hub.py:47]: Do you know if MRL embeddings always follow the `2_Dense_{embedding_dim}/model.safetensors` file format?

**Review by arnavgarg1 — APPROVED** [review]
LGTM

File: server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py (hunk 5 above)
Comment (arnavgarg1) [reply server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py:516]: I see, that's interesting

File: server/lorax_server/utils/sources/hub.py (hunk 7 above)
Comment (magdyksaleh) [reply server/lorax_server/utils/sources/hub.py:47]: for all sentence transformers yes

**Pull request merged**